### PR TITLE
=rem #19780 discard in-flight ACKs for wrong system

### PR DESCRIPTION
**Pull request by patriknw**
- fixes issues noticed as
  "Error encountered while processing system message acknowledgement buffer: [-1 {}] ack: ACK[0, {}]"
  when restarting actor system
- include the origin uid in the ACK and discard received ACKs that
  were targed to an old incarnation of the system with same host:port

### CODE REVIEW

**Comment by patriknw** [comment]
Refs https://github.com/akka/akka/issues/19780

File: akka-cluster/src/multi-jvm/scala/akka/cluster/RestartNodeSpec.scala [hunk 1]
```diff
@@ -27,6 +31,22 @@ object RestartNodeMultiJvmSpec extends MultiNodeConfig {
   commonConfig(debugConfig(on = false).
     withFallback(ConfigFactory.parseString("akka.cluster.auto-down-unreachable-after = 5s")).
     withFallback(MultiNodeClusterSpec.clusterConfig))
+
+  /**
+   * This was used together with sleep in EndpointReader before deliverAndAck
+   * to reproduce issue with misaligned ACKs when restarting system,
```
Comment (ktoso) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/RestartNodeSpec.scala]: Thanks a lot for this comment, I see where that is finally

**Comment by rkuhn** [comment]
LGTM

**Comment by rkuhn** [comment]
Great detective work, @patriknw !

**Comment by ktoso** [comment]
LGTM, great work @patriknw!

**Comment by patriknw** [comment]
let me know if you have any more questions and I can try to explain in a hangout?

**Comment by johanandren** [comment]
Will this for break it for people doing rolling upgrades of their cluster (I know we don't promise to not break)?

**Comment by johanandren** [comment]
NVM saw the comment now!

**Comment by patriknw** [comment]
that's why I added the `originUid` field as optional, so if talking to old system it will be as before

**Comment by johanandren** [comment]
LGTM

**Comment by ktoso** [comment]
Needs a rebase now

**Comment by drewhk** [comment]
LGTM, but why should the ACK carry the UID? Shouldn't be the UID available after the handshake finished? I.e. you should probably just drop the ACKs until the UID of the remote becomes known again? (the ACKs will be resent anyway).

**Comment by patriknw** [comment]
@drewhk the problem was that we got an old ACK (arrived after `GotUid`) that was targeted for the previous incarnation of the system. Can be reproduced as described in the [issue](https://github.com/akka/akka/issues/19780#issuecomment-196280119)

**Comment by drewhk** [comment]
Hm, but how could that happen? I saw the reproduction steps, but the Ack arrives on a TCP connection, so if it arrives from an old connection that means that stopping reading from the previous connection has not been done? I ask this because the underlying issue can make other messages to arrive interleaved with the new ones, not just an Ack.

**Comment by patriknw** [comment]
Good question. I can try to debug it more, since I have a reproducer. It's not that easy to follow what is going on here, as we know.

**Comment by drewhk** [comment]
I can take a look, if you want. With a reproducer it should be at least easier.

**Comment by patriknw** [comment]
I'll like to do it, at least until I'm stuck, thanks

**Comment by drewhk** [comment]
great, I won't complain :D

**Comment by patriknw** [comment]
@drewhk I found the reason and have a better solution. This PR is superseded by https://github.com/akka/akka/pull/20093

**Comment by drewhk** [comment]
Great, you are my hero :)

**Comment by patriknw** [comment]
thanks for pushing me in the right direction

**Pull request closed without merging**